In [86]:
import pandas as pd
import numpy as np
from datetime import datetime

In [87]:
given_data = pd.read_csv('/content/cars_rawdata.csv')

In [88]:
given_data.head()

Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage   Engine  \
0              41000    Diesel       Manual      First  19.67 kmpl  1582 CC   
1              46000    Petrol       Manual      First    13 km/kg  1199 CC   
2              87000    Diesel       Manual      First  20.77 kmpl  1248 CC   
3              40670    Diesel    Automatic     Second   15.2 kmpl  1968 CC   
4              86999    Diesel       Manual      First  23.08 kmpl  1461 CC   

       Power  Seats  New_Price  Price  
0  126.2 bhp    5.0        NaN  12.50  
1   88.7 bhp    5.0  8.61 Lakh   4.50  
2  88.76 bhp    7.0        NaN   6.00  
3  140.8 bhp    5.0        NaN  17.74  
4   63.1 bhp    5.0        NaN   3.50

Replacing units in the columns and handling missing data

In [89]:
# Check for Last values in each column
missed_values= given_data.isnull().sum()
print("last Values:")
print(missed_values)

last Values:
Unnamed: 0              0
Name                    0
Location                0
Year                    0
Kilometers_Driven       0
Fuel_Type               0
Transmission            0
Owner_Type              0
Mileage                 2
Engine                 36
Power                  36
Seats                  38
New_Price            5032
Price                   0
dtype: int64


#### **Mileage, Engine, Power, Seats, and New Price have null or missing data based on the aforementioned observation..**

### **1. Handling Mileage column**

The error shows sample('19.67 kmpl') cannot convert to numeric.

**Solution**: Remove the units.

In [90]:
# Check the initial count of missing values in the 'Mileage' column
print("Initial count of missing values in the Mileage column:", given_data['Mileage'].isnull().sum())

# Use regular expressions to extract numerical values from the 'Mileage' column
# Convert the extracted values to float and replace the existing 'Mileage' column
given_data['Mileage'] = given_data['Mileage'].str.extract('(\d+\.\d+|\d+)').astype(float)

# Since mileage is reported in kmpl, missing values are likely random.
# Fill missing values with the median of the non-missing values
given_data['Mileage'].fillna(given_data['Mileage'].median(), inplace=True)

# Check the final count of missing values in the 'Mileage' column after handling
print("Final count of missing values in the Mileage column after handling:", given_data['Mileage'].isnull().sum())

Initial count of missing values in the Mileage column: 2
Final count of missing values in the Mileage column after handling: 0


### **2. Manage the Engine column**

The error shows sample('300 CC') cannot convert to numeric.

**Solution**: Remove the units.

In [91]:
# Evaluate the initial count of missing values in the 'Engine' column
print("Initial count of missing values in the Engine column:", given_data['Engine'].isnull().sum())

# Extract numerical values from the 'Engine' column and convert them to float
given_data['Engine'] = given_data['Engine'].str.extract('(\d+\.\d+|\d+)').astype(float)

# Handling missing values for engine size often involves filling them with the mean value
# Fill missing values in the 'Engine' column with the mean of the existing non-missing values
given_data['Engine'].fillna(given_data['Engine'].mean(), inplace=True)

# Check the final count of missing values in the 'Engine' column after handling
print("Final count of missing values in the Engine column after handling:", given_data['Engine'].isnull().sum())

Initial count of missing values in the Engine column: 36
Final count of missing values in the Engine column after handling: 0


### **3. Manage the Power Column**

The error shows sample('39.09 bhp') cannot convert to numeric.

**Solution**: Remove the units.

In [92]:
# Evaluate the initial count of missing values in the 'Power' column
print("Initial count of missing values in the Power column:", given_data['Power'].isnull().sum())

# Extract numerical values from the 'Power' column and convert them to float
given_data['Power'] = given_data['Power'].str.extract('(\d+\.\d+|\d+)').astype(float)

# Handling missing values in the 'Power' column is commonly done by filling them with the mean value
# Fill the missing values in the 'Power' column with the mean of the existing non-missing values
given_data['Power'].fillna(given_data['Power'].mean(), inplace=True)

# Check the final count of missing values in the 'Power' column after handling
print("Final count of missing values in the Power column after handling:", given_data['Power'].isnull().sum())

Initial count of missing values in the Power column: 36
Final count of missing values in the Power column after handling: 0


### **4. Manage the seat Column**

In [93]:
# Evaluate the initial count of missing values in the 'Seats' column
print("Initial count of missing values in the Seats column:", given_data['Seats'].isnull().sum())

# It's common practice to address missing values in the 'Seats' column by replacing them with the mean value
# Replace missing values in the 'Seats' column with the mean of the existing non-missing values
given_data['Seats'].fillna(given_data['Seats'].mean(), inplace=True)

# Check the final count of missing values in the 'Seats' column after handling
print("Final count of missing values in the Seats column after handling:", given_data['Seats'].isnull().sum())

Initial count of missing values in the Seats column: 38
Final count of missing values in the Seats column after handling: 0


### **5. manage New_Price Column**

In [94]:
# The New Price column should be removed because there are a lot of missing numbers in it.
given_data.drop('New_Price', axis=1, inplace=True)

**View the no null values handled DataFrame.**

In [95]:
missing_values = given_data.isnull().sum()
missing_values

Unnamed: 0           0
Name                 0
Location             0
Year                 0
Kilometers_Driven    0
Fuel_Type            0
Transmission         0
Owner_Type           0
Mileage              0
Engine               0
Power                0
Seats                0
Price                0
dtype: int64

# `(c)  Convert the categorical variables "Transmission" and "Fuel Type" to a single hot encoded value in numbers.


In [96]:
# Categorical variable encoding using Pandas' get_dummies function
given_data = pd.get_dummies(given_data, columns=['Fuel_Type', 'Transmission'])

# Modified dataset
print(given_data.head())

   Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0              41000      First    19.67  1582.0  126.20    5.0  12.50   
1              46000      First    13.00  1199.0   88.70    5.0   4.50   
2              87000      First    20.77  1248.0   88.76    7.0   6.00   
3              40670     Second    15.20  1968.0  140.80    5.0  17.74   
4              86999      First    23.08  1461.0   63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0                 1                   0                 0   
1                 

# `(d)  Add a new feature to the dataset by creating a new column.

### **added Years_in_service feature which describe about the age of the car.**

In [97]:
# Calculate current year
current_year = datetime.now().year

# To get the age of the car, subtract the current year from the "Year" column.
given_data['Years_in_Service'] = current_year - given_data['Year']

# modified dataset with the new feature
print(given_data.head())

   Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0              41000      First    19.67  1582.0  126.20    5.0  12.50   
1              46000      First    13.00  1199.0   88.70    5.0   4.50   
2              87000      First    20.77  1248.0   88.76    7.0   6.00   
3              40670     Second    15.20  1968.0  140.80    5.0  17.74   
4              86999      First    23.08  1461.0   63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0                 1                   0                 0   
1                 

#### **added Mileage_Category feature to describe mileage value by category.**

In [98]:
# Define the mileage categories
mileage_bins = [0, 10, 20, 30, float('inf')]
mileage_labels = ['Very Low', 'Low', 'Medium', 'High']

# Create a new feature 'Mileage_Category' based on the mileage groups
given_data['Mileage_Category'] = pd.cut(given_data['Mileage'], bins=mileage_bins, labels=mileage_labels, right=False)

# Display the modified dataset with the new feature
print(given_data.head())


   Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0              41000      First    19.67  1582.0  126.20    5.0  12.50   
1              46000      First    13.00  1199.0   88.70    5.0   4.50   
2              87000      First    20.77  1248.0   88.76    7.0   6.00   
3              40670     Second    15.20  1968.0  140.80    5.0  17.74   
4              86999      First    23.08  1461.0   63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0                 1                   0                 0   
1                 

# `(e) Apply the following procedures to this dataset: select, filter, rename, mutate, arrange, and summarize with group by
`

In [99]:
# Selecting required columns
selected_data = given_data[['Name', 'Location', 'Year', 'Mileage', 'Price', 'Years_in_Service']]
# Displaying the results
print("Selected Data:")
print(selected_data.head())

Selected Data:
                               Name    Location  Year  Mileage  Price  \
0  Hyundai Creta 1.6 CRDi SX Option        Pune  2015    19.67  12.50   
1                      Honda Jazz V     Chennai  2011    13.00   4.50   
2                 Maruti Ertiga VDI     Chennai  2012    20.77   6.00   
3   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013    15.20  17.74   
4            Nissan Micra Diesel XV      Jaipur  2013    23.08   3.50   

   Years_in_Service  
0                 9  
1                13  
2                12  
3                11  
4                11  


In [100]:
# Filtering data on yearly basis
filtered_data = given_data[given_data['Year'] >= 2017]
print("\nFiltered Data (Year >= 2019):")
print(filtered_data.head())


Filtered Data (Year >= 2019):
    Unnamed: 0                                     Name Location  Year  \
8           10                         Maruti Ciaz Zeta    Kochi  2018   
15          17                    Maruti Swift DDiS VDI   Jaipur  2017   
26          28                      Honda WRV i-VTEC VX    Kochi  2018   
34          36            Maruti Vitara Brezza ZDi Plus    Kochi  2018   
35          37  Mahindra KUV 100 mFALCON G80 K6 5str AW    Delhi  2017   

    Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
8               25692      First    21.56  1462.0  103.25    5.0   9.95   
15              25000      First    28.40  1248.0   74.00    5.0   5.99   
26              37430      First    17.50  1199.0   88.70    5.0   9.90   
34              50075      First    24.30  1248.0   88.50    5.0   8.63   
35              52000      First    18.15  1198.0   82.00    5.0   4.85   

    Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
8          

In [101]:
# Renaming columns
renamed_data = given_data.rename(columns={'Engine': 'Engine_Displacement', 'Power': 'Horsepower'})

print("\nRenamed Columns:")
print(renamed_data.head())


Renamed Columns:
   Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine_Displacement  Horsepower  \
0              41000      First    19.67               1582.0      126.20   
1              46000      First    13.00               1199.0       88.70   
2              87000      First    20.77               1248.0       88.76   
3              40670     Second    15.20               1968.0      140.80   
4              86999      First    23.08               1461.0       63.10   

   Seats  Price  Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0    5.0  12.50              

In [102]:
# Mutating data (creating new columns)
given_data['Price_Category'] = pd.cut(given_data['Price'], bins=[0, 5, 10, 15, 20, 25, 30], labels=['<5', '5-10', '10-15', '15-20', '20-25', '25-30'])

print("\nMutated Data (Added Price_Category column):")
print(given_data.head())


Mutated Data (Added Price_Category column):
   Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0              41000      First    19.67  1582.0  126.20    5.0  12.50   
1              46000      First    13.00  1199.0   88.70    5.0   4.50   
2              87000      First    20.77  1248.0   88.76    7.0   6.00   
3              40670     Second    15.20  1968.0  140.80    5.0  17.74   
4              86999      First    23.08  1461.0   63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0                 1               

In [103]:
# Data arranged in descending order according to year
arranged_data = given_data.sort_values(by='Year', ascending=False)

print("\nArranged Data (Sorted yearly basis data in descending order):")
print(arranged_data.head())


Arranged Data (Sorted yearly basis data in descending order):
      Unnamed: 0                             Name    Location  Year  \
5405        5560        Renault KWID RXT Optional       Kochi  2019   
942          975  Ford Endeavour 2.2 Trend AT 4X2       Kochi  2019   
5533        5690        Maruti Omni 5 Seater BSIV  Coimbatore  2019   
770          796            Mahindra XUV500 W9 AT  Coimbatore  2019   
4267        4399       Maruti Swift Dzire AMT ZDI     Chennai  2019   

      Kilometers_Driven Owner_Type  Mileage  Engine  Power  Seats  Price  \
5405               6568      First    25.17   799.0   53.3    5.0   5.09   
942               11209      First    12.62  2198.0  158.0    7.0  31.15   
5533               4721      First    14.00   796.0   35.0    5.0   4.11   
770               19654      First    14.00  2179.0  155.0    7.0  17.63   
4267              65000      First    26.59  1248.0   74.0    5.0   6.75   

      Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type

In [104]:
# Summarized data with group by operations
summary_data = given_data.groupby('Location')['Price'].mean()

print("\nSummary Data (Average Price by Location):")
print(summary_data)


Summary Data (Average Price by Location):
Location
Ahmedabad      8.567248
Bangalore     13.482670
Chennai        7.958340
Coimbatore    15.160206
Delhi          9.881944
Hyderabad      9.997423
Jaipur         5.916725
Kochi         11.309109
Kolkata        5.733924
Mumbai         9.592546
Pune           6.951000
Name: Price, dtype: float64


# ` given_data(DataFrame) in  CSV`

In [105]:
given_data.head()

Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0              41000      First    19.67  1582.0  126.20    5.0  12.50   
1              46000      First    13.00  1199.0   88.70    5.0   4.50   
2              87000      First    20.77  1248.0   88.76    7.0   6.00   
3              40670     Second    15.20  1968.0  140.80    5.0  17.74   
4              86999      First    23.08  1461.0   63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0                 1                   0                 0   
1                 0                   0                 1   
2                 1                   0                 0   
3                 1                   0                 0   
4                 1                   0                 0   

   Transmission_Automatic  Transmission_Manual  Years_in_Service  \
0                       0                    1                 9   
1                       0                    1                13   
2                       0                    1                12   
3                       1                    0                11   
4                       0                    1                11   

  Mileage_Category Price_Category  
0              Low          10-15  
1              Low             <5  
2           Medium           5-10  
3              Low          15-20  
4           Medium             <5

In [106]:
#Save the data into clean data csv
given_data.to_csv("/content/clean_data.csv")